Inventory retry.

merlin-northern opened the pull request:
We use RetryPollIntervalSeconds in inventory with the exponential
backoff via GetExponentialBackoffTime together with a new setting:
* RetryPollCount -- the max number of tries
* inventory by default tries 3 times with one minute intervals
  (GetExponentialBackoffTime defaults)

ChangeLog:commit
ChangeLog:RetryPollCount applies to all places where backoff is present
Signed-off-by: Peter Grzybowski <peter@northern.tech>

merlin-northern commented:
@mender-test-bot start pipeline please

merlin-northern commented:
@mender-test-bot start pipeline sugar pretty please

merlin-northern commented:
@mender-test-bot start pipeline sugar pretty pretty please

merlin-northern commented:
@kacf I am asking review fyi here (wrt the private discussion)

kacf commented on app/state.go:
```diff
@@ -1104,6 +1109,68 @@ func (fir *fetchStoreRetryState) Handle(ctx *StateContext, c Controller) (State,
+}
+func (fir *inventoryUpdateRetry) Handle(ctx *StateContext, c Controller) (State, bool) {
+	if ctx.nextAttemptAt.After(time.Now()) {
+		secondsRemaining := time.Duration(ctx.nextAttemptAt.Unix()-time.Now().Unix()) * time.Second
```
Should use: `secondsRemaining := ctx.nextAttemptAt.Sub(time.Now()) * time.Second`

merlin-northern replied:
corrected. (dropped the time.Second() as Sub returns duration)

merlin-northern replied:
and the static checker was smarter -- suggested Until

kacf commented on app/state.go:
```diff
@@ -1104,6 +1109,68 @@ func (fir *fetchStoreRetryState) Handle(ctx *StateContext, c Controller) (State,
+		return fir.Wait(NewInventoryUpdateState(), fir, secondsRemaining, ctx.WakeupChan)
+	}
+	log.Infof("Handle update inventory retry state try: %d", ctx.inventoryUpdateAttempts)
+	err := c.InventoryRefresh()
+	if err != nil {
+		log.Warnf("Failed to refresh inventory: %v", err)
+		if errors.Cause(err) == errNoArtifactName {
+			return NewErrorState(NewTransientError(err)), false
+		}
+	}
+
+	interval, err := client.GetExponentialBackoffTime(
+		ctx.inventoryUpdateAttempts,
+		c.GetUpdatePollInterval(),
+		ctx.retryCount,
+	)
+	ctx.nextAttemptAt = time.Now().Add(interval)
+	if err != nil {
+		log.Infof("Handle update inventory retry state: failed to send inventory: %s", err.Error())
+		return States.CheckWait, false
+	}
+
+	ctx.inventoryUpdateAttempts++
+
+	if interval.Seconds() > ctx.inventoryPollInterval.Seconds() {
+		interval = ctx.inventoryPollInterval
+	}
+	log.Infof("Wait %v before next inventory update attempt", interval)
+	return fir.Wait(NewInventoryUpdateState(), fir, interval, ctx.WakeupChan)
```
As far as I can see, there is no success condition here. It will always wait, regardless of whether the refresh worked or not.

merlin-northern replied:
right. that was really a preliminary call for a review. thanks!

kacf commented on app/state.go:
```diff
@@ -1194,6 +1261,15 @@ type inventoryUpdateState struct {
 	baseState
 }
 
+func NewInventoryUpdateState() State {
```
Question: Since this is a new function, but the `inventoryUpdateState` struct is not new, where is this constructed in the existing code? We should probably change all instances to use this function now that we have it.

merlin-northern replied:
right. I have found only one: https://github.com/mendersoftware/mender/pull/943/commits/f27eaffbdde042b3c89b8f1c83e797d3bc08417a#diff-0b03a1a30843453d28f38533c95445ff2cd5009a25d18c640ef66426efc044e8R93

merlin-northern commented:
@mender-test-bot start pipeline sugar really pretty please

kacf commented on app/state_test.go:
```diff
@@ -494,7 +499,7 @@ func TestStateInvetoryUpdate(t *testing.T) {
 		inventoryErr: errors.New("some err"),
 	})
-	assert.IsType(t, &checkWaitState{}, s)
+	assert.IsType(t, &inventoryUpdateRetry{}, s)
```
Would be great to test the new wait state as well, which should get the coverage up.

merlin-northern replied:
added! https://github.com/mendersoftware/mender/pull/943/commits/383fca9e42d0f52253be1f872e48554e6585ce97#diff-ef52b340986b4690b1936c29f3f7e22bd4df65857dad63ffd279e72b40838d8fR515

and in passing also removed the unneeded argument here: https://github.com/mendersoftware/mender/compare/d21fe3d1b38b29df6d91796a3865bd3e3cac47ab..7d9bae4484013fc2c7f1baaa60a02b5d1c772504#diff-0b03a1a30843453d28f38533c95445ff2cd5009a25d18c640ef66426efc044e8R1114

kacf reviewed (approved):
I think this should have a unit test for the new state, otherwise the code looks good, so pre-approving.

merlin-northern commented:
@mender-test-bot start pipeline sugar really pretty please with cherry on top

lluiscampos commented:
@merlin-northern Great changelog entries. Thank you :rocket:

merlin-northern commented:
> @merlin-northern Great changelog entries. Thank you 🚀

@kacf fault! he told me to make two of them :)

merlin-northern commented:
@kacf I could not stop.
I added a new [commit](https://github.com/mendersoftware/mender/pull/943/commits/e9cebd419a90272b654fc9f10b05de008a2a9cd1), fixing a couple of things, I will squash it if approved
1. I removed 
```diff
-			retryCount:            mender.GetRetryPollCount(),
-			inventoryPollInterval: mender.GetInventoryPollInterval(),
```
since we can rely on the controller for those
2. we are sleeping in the inventory retry state with the smallest configured poll inverval, this is to cover a case when there is very long inventory retry timeout and much shorter check update interval -- we should not delay anything.
3. the rest are tests adjustments.

merlin-northern commented:
@mender-test-bot start pipeline sweet please

merlin-northern commented:
@mender-test-bot start pipeline sweet please

merlin-northern commented:
@mender-test-bot start pipeline sweet pretty please with cherry

merlin-northern commented:
@mender-test-bot start pipeline please; dont panic

merlin-northern commented:
docs update: https://github.com/mendersoftware/mender-docs/pull/1700

merlin-northern commented:
as much as I would love to add acceptance tests for the inventory update, the effort in comparison seems too big. we need to use and configure the mock server as well as write the extra tests. as this is unit tested and all other tests pass, I decide to merge it with decreasing the acceptance tests coverage. it was discussed.
